### ISSUE TITLE
Publishing non-SNAPSHOT release?

### ISSUE THREAD

**Issue by mmoayyed** [codecentric/spring-boot-admin#1268]
First, thanks very much for this great tool.

I have been testing the 2.2.0-SNAPSHOT as part of the [Apereo CAS software](https://github.com/apereo/cas) for a while and from time to time, we run into issues like this:

```
* What went wrong:
Execution failed for task ':webapp:cas-server-webapp-init-bootadmin-server:compileJava'.
> Could not resolve all files for configuration ':webapp:cas-server-webapp-init-bootadmin-server:compileClasspath'.
   > Could not find de.codecentric:spring-boot-admin-server:2.2.0-SNAPSHOT.
```

Could you please consider publishing a non-snapshot release so that we may avoid brittle issues with builds? Sonatype has been quite fragile over the past few days, and in general, it should help folks and eager testers better if they had a milestone/release-candidate to work and experiment with. 

Thanks again.

**Comment by joshiste** [comment]
Sorry but I can't do this:

1) Sonatype provides no milestone repository I could publish to.
2) We can't publish to Maven Central with depending on Milestones of Spring Boot / Spring Cloud

**Comment by mmoayyed** [comment]
Thanks for the update. Are you saying sonatype disallows you from doing this, or this is something you can't do for this projects? 

I understand the issue and appreciate your response. Just wanted to clarify that we publish release candidates and milestones releases all the time to sonatype repos while depending on non-ga versions of many dependencies; I understand if that is not a favorable approach for you but wanted to clarify that this is technically possible. 

Thanks again.

**Comment by joshiste** [comment]
> Just wanted to clarify that we publish release candidates and milestones releases all the time to sonatype

to your private instance? I'm talking about https://oss.sonatype.org

**Comment by mmoayyed** [comment]
> to your private instance? I'm talking about oss.sonatype.org

Not a private instance. Yes the very same `oss.sonatype.org` URL.

- [Example: Release Candidate](https://search.maven.org/artifact/org.apereo.cas/cas-server-core/6.1.0-RC5/jar)
- [Example: Milestone Release](https://search.maven.org/artifact/org.apereo.cas/cas-server-core/5.0.0.M3/jar)

In either of those releases, you'll find SNAPSHOT dependencies in pom.xml files. All non-SNAPSHOT releases are published to https://oss.sonatype.org/service/local/staging/deploy/maven2/.

The requirement with Sonatype for that particular repository, IIRC, is that the version number, a mere string, must not end with `SNAPSHOT`. No other deep analysis is performed to figure out the exact meaning and nature of a given release. So, you could, for instance, publish `2.2.0-M1`, `2.2.0-ReleaseCandidateIzGr8`, or `2.2.0-Angry-Banana` :) All should be fine, because from Sonatype's perspective they are non-SNAPSHOTs. 

HTH. 

PS Side note: I am personally considering a move from Sonatype to Bintray. Sonatype OSSRH has had [quite the week](https://status.maven.org/) which I think exasperated the use case here.

**Comment by joshiste** [comment]
according to the [OSSRH Guide](https://central.sonatype.org/pages/ossrh-guide.html) the staging repositories have no download access, so I think this won't do.

**Comment by mmoayyed** [comment]
> According to the OSSRH Guide the staging repositories have no download access, so I think this won't do.

Not only is that unnecessary, it is specially unnecessary for this task. If I understand your comment correctly, direct download access is not needed and isn't useful, because one wouldn't want to download an artifact directly from that URL, just like one wouldn't download a snapshot artifact from the snapshot repositories or URLs. The process is no different than cutting *an actual* GA release. Once you publish a release candidate or milestone, (as you would normally any other type of release), you'd log into oss.sonatype.org, close the repository and the publish the release. The release is then synced with maven central and would then be available for download either directly from central, or more adequately as most would do, via a build tool such as Maven or Gradle.

I hope that clarifies better.

**Comment by joshiste** [comment]
> close the repository and the publish the release

wouldn't that mean in the end the milestone would be synced to maven central?

**Comment by mmoayyed** [comment]
Yes certainly. Would that cause issues?

**Comment by joshiste** [comment]
Yeah. IIRC you should only publish things, which dependencies are available on maven central.
That is not the case for the Milestones of Spring / Spring Boot / Spring Cloud - so we can't publish milestones to maven central.
